rumqttd/architecture.md

```diff
@@ -139,5 +139,97 @@ NOTE: Transition to ConnectionReady will schedule the connection for consumption
+
+   An action of any sort on the router's end which involves communication usually involves the router adding a structure or packet to one of these buffers.
+
+   The router also shares Channels, which do not contain the events themselves, but rather contain notifications to events that might be added into the shared event buffers. This also has the added benefit of preventing lock contention between different futures/tasks, which would've happened if they had to keep checking the buffers
+
+   When you create the router, the following happens ->
+
+   1. Transmission Channels get created to and from the router
+   2. A `Graveyard` to store information to persist connections
+   3. A tracker for All the alerts and metrics that are necessary to be handled
+   4. A Map between subscriptions and devices gets created
+   5. The two communication buffers `ibufs` and `obufs` get initialized
+   6. Logs for message acknowledgements, subscriptions, and alerts get created
+   7. A Packet Cache is created
+   8. The Scheduler struct is created
+
+   Once all necessary things have been initialized, the `spawn` method intializes the router's event loop on a separate thread.
+   This method returns a cloneable transmission channel that the Broker passes to all links so they can communicate with the router.
+
+3. Link
+
+   An Asynchronous and usually network-facing entity that Handles transmission of information to and from the Router. Information here can mean things including but not limited to
+
+   1. Messages incoming from devices (publish)
+   2. Messages outgoing to devices (subscribe)
+   3. Information about metrics
+   4. Other events such as pings
+
+   There are various types of Links. Here they are, explored in minimal detail -
+
+   1. Bridge -> Can be used to link two `rumqttd` instances for high availability.
+   2. Alerts -> Since `rumqttd` is embeddable, this allows developers to create custom hooks to handle alerts that `rumqttd` gives out.
+   3. Console -> Creates a lightweight HTTP Server to obtain information and metrics from the `rumqttd` instance.
+   4. Local -> Contains functions to perform Publish and Subscribe actions to the router
+   5. Meters -> Creates a link specifically to consume metrics, is useful for external observability or hooks
+   6. Network -> Contains the APIs for reading from and writing to sockets
+   7. Remote -> `RemoteLink` is an asynchronous entity that represents a network connection. It provides its own event loop, which does the following -
+      - Reads from the network socket, writes to `ibufs` and notifies the router
+      - Writes to the network socket when required
+   8. Shadow -> TODO
+   9. Timer -> Handles timing actions for metrics
```

Comment:
it sends `Events` related to alerts and meters which causes router to sends alerts and meters to their respective links.

can think of it as required link for `alerts` and `meters`